File: lintcode/python/1783_binary_tree_postorder_traversal_null.py
#!/usr/bin/python -t

# recursive traversal solution

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: A Tree
    @return: Postorder in ArrayList which contains node values.
    """
    def __init__(self):
        self.ret = []
        
    def postorderTraversal(self, root):
        # write your code here
        self.traverse(root)
        
        return self.ret
        
    def traverse(self, node):
        if node == None:
            return
        
        self.traverse(node.left)
        self.traverse(node.right)
        self.ret.append(node.val)
        
        return
    

File: lintcode/golang/1783_binary_tree_postorder_traversal_null.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

/**
 * @param root: A Tree
 * @return: Postorder in ArrayList which contains node values.
 */
func postorderTraversal (root *TreeNode) []int {
    // write your code here
    if root == nil {
        return []int{}
    }
    
    var ret []int
    var helper func(*TreeNode, *[]int)
    helper = func(node *TreeNode, l *[]int) {
        if node == nil {
            return
        }
        helper(node.Left, l)
        helper(node.Right, l)
        *l = append(*l, node.Val)
        return
    }
    
    helper(root, &ret)
    
    return ret
}

